File: src/domain/formatMoney.ts
import { assertHalalas, minorUnitsOf, type Currency, type Halalas } from './money'

/**
 * تنسيق المبالغ للعرض — **المكان الوحيد** الذي تتحول فيه الهللة إلى ريال.
 *
 * التحويل هنا نصّي بحت (قسمة صحيحة + باقي)، لا قسمة عشرية،
 * فلا يظهر 96.46999999 أبدًا.
 *
 * ملاحظة معمارية: الملف في domain/ لأنه دالة نقية بلا اعتماد خارجي،
 * لكن استعماله في presentation/ فقط.
 */

export interface FormatOptions {
  /** إظهار رمز العملة. الافتراضي: نعم */
  showCurrency?: boolean
  /** إظهار إشارة + للمبالغ الموجبة (حركة الأموال). الافتراضي: لا */
  alwaysSign?: boolean
  /** فاصل الآلاف. الافتراضي: نعم */
  grouping?: boolean
  /** إخفاء المبالغ (إعداد الخصوصية في spec/01). */
  hidden?: boolean
}

const CURRENCY_LABEL: Record<Currency, string> = {
  SAR: 'ر.س',
  EGP: 'ج.م',
  USD: '$',
  EUR: '€',
  GBP: '£',
  AED: 'د.إ',
}

export function currencyLabel(currency: Currency): string {
  return CURRENCY_LABEL[currency] ?? currency
}

function groupThousands(digits: string): string {
  let out = ''
  for (let i = 0; i < digits.length; i++) {
    const fromEnd = digits.length - i
    out += digits[i]
    if (fromEnd > 1 && (fromEnd - 1) % 3 === 0) out += ','
  }
  return out
}

/**
 * يحوّل الهللة إلى نص بالريال. بدون رمز العملة.
 * 9647 → "96.47" · -9647 → "-96.47" · 0 → "0.00" · 100000 → "1,000.00"
 */
export function formatAmount(
  amount: Halalas,
  currency: Currency = 'SAR',
  options: FormatOptions = {},
): string {
  assertHalalas(amount, 'مبلغ للعرض')
  const { alwaysSign = false, grouping = true } = options

  const scale = minorUnitsOf(currency)
  const decimals = String(scale).length - 1

  const negative = amount < 0
  const abs = Math.abs(amount)

  const whole = Math.floor(abs / scale)
  const frac = abs - whole * scale

  const wholeStr = grouping ? groupThousands(String(whole)) : String(whole)
  const fracStr = decimals > 0 ? '.' + String(frac).padStart(decimals, '0') : ''

  const sign = negative ? '-' : alwaysSign ? '+' : ''
  return sign + wholeStr + fracStr
}

/** نص كامل للعرض مع رمز العملة: "96.47 ر.س". */
export function formatMoney(
  amount: Halalas,
  currency: Currency = 'SAR',
  options: FormatOptions = {},
): string {
  const { showCurrency = true, hidden = false } = options
  if (hidden) return showCurrency ? `•••• ${currencyLabel(currency)}` : '••••'
  const num = formatAmount(amount, currency, options)
  return showCurrency ? `${num} ${currencyLabel(currency)}` : num
}

/**
 * القيمة غير المتاحة تُعرض كنص صريح، لا صفر.
 * قاعدة CLAUDE.md #10: لا رقم بلا مصدر، ولا صفر مؤكد مكان المجهول.
 */
export const NOT_AVAILABLE = 'غير متاح'

export function formatMoneyOrNA(
  amount: Halalas | null | undefined,
  currency: Currency = 'SAR',
  options: FormatOptions = {},
): string {
  if (amount === null || amount === undefined) return NOT_AVAILABLE
  return formatMoney(amount, currency, options)
}

/**
 * معدل الادخار كنسبة مئوية بخانة عشرية واحدة، أو null لو الدخل صفر.
 * spec/06: «دخل0 → الادخار غير متاح؛ لا Infinity/NaN». والسالب لا يُخفى.
 */
export function savingsRatePercent(incomeMinor: Halalas, remainingMinor: Halalas): number | null {
  assertHalalas(incomeMinor, 'الدخل')
  assertHalalas(remainingMinor, 'المتبقي')
  if (incomeMinor === 0) return null
  // بالعُشر من المئة ثم قسمة على 10 عند العرض فقط
  const tenths = Math.round((remainingMinor * 1000) / incomeMinor)
  return tenths / 10
}

export function formatPercentOrNA(percent: number | null): string {
  if (percent === null || !Number.isFinite(percent)) return NOT_AVAILABLE
  const sign = percent < 0 ? '-' : ''
  const abs = Math.abs(percent)
  return `${sign}${abs.toFixed(1)}%`
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/FormatMoney.kt
package app.masroufy.core

import kotlin.math.abs

/**
 * عرض المبالغ — نقل `src/domain/formatMoney.ts`. **المكان الوحيد** اللي الهللة بتبقى ريال فيه،
 * بقسمة صحيحة وباقي (مفيش 96.46999 أبدًا).
 */

private fun currencyLabel(currency: Currency): String = when (currency) {
    Currency.SAR -> "ر.س"
    Currency.EGP -> "ج.م"
    Currency.USD -> "$"
    Currency.EUR -> "€"
    Currency.GBP -> "£"
    Currency.AED -> "د.إ"
}

private fun groupThousands(digits: String): String {
    val out = StringBuilder()
    for (i in digits.indices) {
        val fromEnd = digits.length - i
        out.append(digits[i])
        if (fromEnd > 1 && (fromEnd - 1) % 3 == 0) out.append(',')
    }
    return out.toString()
}

/** 9647 ⇒ "96.47" · -9647 ⇒ "-96.47" · 0 ⇒ "0.00" · 100000 ⇒ "1,000.00" — من غير رمز العملة. */
fun formatAmount(
    amount: Halalas,
    currency: Currency = Currency.SAR,
    alwaysSign: Boolean = false,
    grouping: Boolean = true,
): String {
    assertHalalas(amount, "مبلغ للعرض")
    val scale = minorUnitsOf(currency)
    val decimals = scale.toString().length - 1
    val total = abs(amount)
    val whole = total / scale
    val frac = total - whole * scale
    val wholeText = if (grouping) groupThousands(whole.toString()) else whole.toString()
    val fracText = if (decimals > 0) "." + frac.toString().padStart(decimals, '0') else ""
    val sign = if (amount < 0) "-" else if (alwaysSign) "+" else ""
    return sign + wholeText + fracText
}

/** نص كامل برمز العملة: "96.47 ر.س". `hidden` = إخفاء المبالغ (spec/01). */
fun formatMoney(
    amount: Halalas,
    currency: Currency = Currency.SAR,
    showCurrency: Boolean = true,
    alwaysSign: Boolean = false,
    grouping: Boolean = true,
    hidden: Boolean = false,
): String {
    if (hidden) return if (showCurrency) "•••• ${currencyLabel(currency)}" else "••••"
    val number = formatAmount(amount, currency, alwaysSign, grouping)
    return if (showCurrency) "$number ${currencyLabel(currency)}" else number
}

/** القيمة المجهولة نص صريح مش صفر (CLAUDE.md #10). */
const val NOT_AVAILABLE = "غير متاح"

fun formatMoneyOrNA(amount: Halalas?, currency: Currency = Currency.SAR): String =
    if (amount == null) NOT_AVAILABLE else formatMoney(amount, currency)

/**
 * معدل الادخار بخانة عشرية واحدة، أو null لو الدخل صفر (spec/06: لا Infinity ولا NaN، والسالب ما يتخباش).
 * نسبة للعرض مش مبلغ — نفس حساب التطبيق الحالي بالظبط (قسمة double ثم تقريب جافاسكربت).
 */
fun savingsRatePercent(incomeMinor: Halalas, remainingMinor: Halalas): Double? {
    assertHalalas(incomeMinor, "الدخل")
    assertHalalas(remainingMinor, "المتبقي")
    if (incomeMinor == 0L) return null
    val tenths = JsText.round((remainingMinor.toDouble() * 1000.0) / incomeMinor.toDouble())
    return tenths / 10.0
}

/** "12.5%" — دقيق للقيم اللي فيها خانة عشرية واحدة (ناتج `savingsRatePercent`). */
fun formatPercentOrNA(percent: Double?): String {
    if (percent == null || !percent.isFinite()) return NOT_AVAILABLE
    val sign = if (percent < 0) "-" else ""
    val tenths = JsText.round(abs(percent) * 10.0).toLong()
    return "$sign${tenths / 10}.${tenths % 10}%"
}
